baseapp/abci.go

```diff
@@ -158,50 +159,94 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // the ante handler (which checks signatures/fees/ValidateBasic).
 //
 // NOTE:CheckTx does not run the actual Msg handler function(s).
-func (app *BaseApp) CheckTx(req abci.RequestCheckTx) (res abci.ResponseCheckTx) {
-	var result sdk.Result
-
+func (app *BaseApp) CheckTx(req abci.RequestCheckTx) abci.ResponseCheckTx {
 	tx, err := app.txDecoder(req.Tx)
+	if err != nil {
+		space, code, log := sdkerrors.ABCIInfo(err, false)
+		return abci.ResponseCheckTx{
+			Codespace: space,
+			Code:      code,
+			Log:       log,
+		}
+	}
+
+	var (
+		space  string
+		code   uint32
+		log    string
+		data   []byte
+		events []abci.Event
+		mode   runTxMode
+	)
+
 	switch {
-	case err != nil:
-		result = err.Result()
 	case req.Type == abci.CheckTxType_New:
-		result = app.runTx(runTxModeCheck, req.Tx, tx)
+		mode = runTxModeCheck
+
 	case req.Type == abci.CheckTxType_Recheck:
-		result = app.runTx(runTxModeReCheck, req.Tx, tx)
+		mode = runTxModeReCheck
+
 	default:
-		panic(fmt.Sprintf("Unknown RequestCheckTx Type: %v", req.Type))
+		panic(fmt.Sprintf("unknown RequestCheckTx type: %s", req.Type))
+	}
+
+	gInfo, result, err := app.runTx(mode, req.Tx, tx)
+	if err != nil {
+		space, code, log = sdkerrors.ABCIInfo(err, false)
+	} else if result != nil {
+		data = result.Data
+		log = result.Log
+		events = result.Events.ToABCIEvents()
 	}
 
 	return abci.ResponseCheckTx{
-		Code:      uint32(result.Code),
-		Data:      result.Data,
-		Log:       result.Log,
-		GasWanted: int64(result.GasWanted), // TODO: Should type accept unsigned ints?
-		GasUsed:   int64(result.GasUsed),   // TODO: Should type accept unsigned ints?
-		Events:    result.Events.ToABCIEvents(),
+		GasWanted: int64(gInfo.GasWanted), // TODO: Should type accept unsigned ints?
```

Comment:
Okay, so the purpose to combine these two is for the gas Wanted/Used?

Maybe update the function I proposed above to take `(err error, gas GasInfo)` and then it will work for the top case (0 gas) and the bottom case. And you can just short-circuit return in the err != nil phase, rather than this slightly-confusing flow control

Reply:
Wasn't a fan of the control flow either. I'll tweak it some more.